File: server/sonar-server/src/main/webapp/WEB-INF/app/models/widget_property.rb
#
# SonarQube, open source software quality management tool.
# Copyright (C) 2008-2014 SonarSource
# mailto:contact AT sonarsource DOT com
#
# SonarQube is free software; you can redistribute it and/or
# modify it under the terms of the GNU Lesser General Public
# License as published by the Free Software Foundation; either
# version 3 of the License, or (at your option) any later version.
#
# SonarQube is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
# Lesser General Public License for more details.
#
# You should have received a copy of the GNU Lesser General Public License
# along with this program; if not, write to the Free Software Foundation,
# Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
#
class WidgetProperty < ActiveRecord::Base
  belongs_to :widget

  validates_length_of :kee, :within => 1..100
  validates_length_of :text_value, :maximum => 4000, :allow_blank => true, :allow_nil => true

  def key
    kee
  end

  def text_value
    read_attribute(:text_value) || default_text_value
  end

  def default_text_value
    java_definition.defaultValue()
  end

  def type
    @type ||=
      begin
        java_definition.type().name()
      end
  end

  def java_definition
    @java_definition ||=
      begin
        widget.java_definition.getWidgetProperty(key)
      end
  end

  def value
    WidgetProperty.text_to_value(text_value, type)
  end

  def to_hash_json
    {:key => key, :value => text_value}
  end

  def to_xml(xml=Builder::XmlMarkup.new(:indent => 0))
    xml.property do
      xml.key(prop_key)
      xml.value { xml.cdata!(text_value) }
    end
    xml
  end

  def self.text_to_value(text, type)
    PropertyType.text_to_value(text, type);
  end

  protected
  def validate
    errors.add_to_base("Unknown property: #{key}") unless java_definition
    PropertyType::validate(key, type, java_definition.optional(), text_value, errors)
  end

end

File: sonar-plugin-api/src/main/java/org/sonar/api/web/WidgetProperty.java
/*
 * SonarQube
 * Copyright (C) 2009-2018 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
package org.sonar.api.web;

import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

/**
 * @deprecated since 6.2, this extension is ignored as dashboards have been removed
 */
@Deprecated
@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.TYPE)
public @interface WidgetProperty {

  String key();

  WidgetPropertyType type() default WidgetPropertyType.STRING;

  String defaultValue() default "";

  String description() default "";

  boolean optional() default true;

  /**
   * @since 3.3  Options for property of type {@link WidgetPropertyType#METRIC}.
   *
   * If no option is specified, any metric will match.
   * If options are specified, all must match for the metric to be displayed.
   * Three types of filter are supported <code>key:REGEXP</code>, <code>domain:REGEXP</code> and <code>type:comma_separated__list_of_types</code>.
   * For example <code>key:new_.*</code> will match any metric which key starts by <code>new_</code>.
   * For example <code>type:INT,FLOAT</code> will match any metric of type <code>INT</code> or <code>FLOAT</code>.
   * For example <code>type:NUMERIC</code> will match any metric of numerictype.
   *
   * @since 3.5  Options for property of type {@link WidgetPropertyType#SINGLE_SELECT_LIST}.
   * For example {"property_1", "property_3", "property_3"}).
   *
   */
  String[] options() default {};
}
